library/ssl_tls13_server.c

```diff
@@ -980,7 +1001,9 @@ MBEDTLS_CHECK_RETURN_CRITICAL
 static int ssl_tls13_check_psk_key_exchange(mbedtls_ssl_context *ssl)
 {
 #if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED)
-    return mbedtls_ssl_conf_tls13_psk_enabled(ssl) &&
+    return ssl_tls13_ticket_permission_check(
```

Comment:
I think the proposed change does not make much diffenrence with current implementation. The main reason of discarding a selected PSK (not only PSKs provisioned via NewSessionTicket, but also externally provisioned PSKs) is we prefer ephemeral over pure psk. It's hard to avoid this kind of discard unless we do key exchange mode determination in `ssl_tls13_parse_pre_shared_key_ext`, but I think it is not optimal either.

BTW, are there any secure considerations that we should not discard a selected PSK?

Reply:
> In the other case, we can end up with an handshake failure because we check the ticket flag here and not in ssl_tls13_parse_pre_shared_key_ext().

I know you point, but we already have ticket flags check in `ssl_tls13_parse_pre_shared_key_ext()`, do we? Please check https://github.com/Mbed-TLS/mbedtls/pull/6616.

https://github.com/Mbed-TLS/mbedtls/blob/51328162e6da548d369a94b71e47a2b5d571f88f/library/ssl_tls13_server.c#L166-L184

> Thus I definitely think that we should move the ticket flags checks in ssl_tls13_parse_pre_shared_key_ext() to avoid selecting a ticket that actually we cannot use for the handshake.

I agree that we need to check `ticket_flags` twice:
1. In `ssl_tls13_parse_pre_shared_key_ext`: to ensure we do not select a ticket that is not compatible with the proposed key exchange mode (e.g. values in `PskKeyExchangeModes` extension).
2. Here in `ssl_tls13_determine_key_exchange_mode` (end of `ssl_tls13_parse_client_hello`): to ensure we do not select a key exchange mode that is not allowed by the selected ticket.

I think `point 1` have been done by https://github.com/Mbed-TLS/mbedtls/pull/6616, and this PR is to achieve `point 2`.

Reply:
The code you refer to above in `ssl_tls13_offered_psks_check_identity_match_ticket()` does indeed some checks of the ticket flags: compatibility with the key exchange modes proposed by the client. I suggest to do also the checks that the ticket is actually usable by the server to proceed the handshake here thus while we are looking for an usable ticket or PSK. That way if the ticket is actually not usable to proceed the handshake, maybe another is.

Reply:
Yes, I think the changes in `f8e50a9` are right and we need them. With these changes, we do not select anymore a ticket that we cannot actually use because of the restrictions imposed by its key exchange mode flags.

Then with these changes in place, I feel the need to refactor the end of `ssl_tls13_parse_client_hello()` (along the lines of my initial proposal, the proposal does probably miss some cases) otherwise when calling `ssl_tls13_determine_key_exchange_mode()` we are going basically to redo the same checks as the ones in `f8e50a9` in PSK and PSK ephemeral case. 
 
But this could be done in a follow-up PR.

In this PR I would like to have `f8e50a9` and the fact that we no longer use a ticket in a key exchange mode where we are not allowed to use it.